// dbServer/operator.js
var mysql = require('mysql');

var fs = require('fs');

var dbServer = {};

var mewutil = require('mew_util');
var async = mewutil.async;

var db_config = {
	host: '127.0.0.1',
	user: 'root',
	password: '6736498jcf',
	port: '3306',
	database: 'ncg'
};

var initSql = function() {
	fs.readFile('whitecloud.sql', function(err, data) {
		Sqls = data.toString().split(';');
		async.all(Sqls, function(sql) {
			if (sql != '' && sql != '\n') {
				dbServer.operation(sql + ';', null);
				this.next();
			}
		})
	})
}

dbServer['init'] = function(callback) {
	dbServer.connection = mysql.createConnection(db_config);
	dbServer.connection.connect(function(err) {
		if (err) {
			console.log("database link error: " + err);
			console.log("database relink " + new Date());
			setTimeout(dbServer.init, 2000); //2秒重连一次
			return;
		}
		console.log("database linked");
		//initSql();
		if (callback) {
			callback();
		}
	});
	dbServer.connection.on('error', function(err) {
		if (err.code === 'PROTOCOL_CONNECTION_LOST') {
			dbServer.init();
		} else {
			throw err;
		}
	});
}

dbServer['end'] = function() {
	dbServer.connection.end(function(err) {
		if (err) {
			console.log('[connection end] erro!');
		} else {
			console.log('[connection end] succeed!')
		}
	})
}

dbServer['operation'] = function(userAddSql, userAddSql_Params, callback) {
	dbServer.connection.query(userAddSql, userAddSql_Params, function(err, result) {
		if (err) {
			if (callback) {
				callback(err, null);
			}
		} else {
			if (callback) {
				callback(null, result);
			};
		}
	});
}

dbServer['escape'] = function(string) {
	return dbServer.connection.escape(string);
}

module.exports = dbServer;
